lib/rucio/core/authentication.py

```diff
@@ -349,17 +446,856 @@ def get_auth_token_saml(account, saml_nameid, appid, ip=None, session=None):
     return new_token
 
 
+def get_rucio_OIDC_clients(keytimeout=43200):
+    """
+    Creates a Rucio OIDC Client instances as defined per Identity Provider (IdP) in the
+    etc/idpsecrets.json configuration file. These have to be pre-registered with the
+    respective IdP with the appropriate settings.
+    (allowed to request refresh tokens, token exchange, immediate refresh tokens expiration
+    after first use, refresh tokens have lifetime set in their unverified header)
+
+    :returns: Dictionary of {'https://issuer_1/': <Rucio OIDC Client_1 instance>,
+              'https://issuer_2/': <Rucio OIDC Client_2 instance>,}.
+              In case of trouble, Exception is raised.
+    """
+    # initializing a client_id and client_secret - provided in a secret config file
+    with open(idpsecrets) as client_secret_file:
+        client_secrets = json.load(client_secret_file)
+    clients = {}
+    for iss in client_secrets:
+        try:
+            client_secret = client_secrets[iss]
+            issuer = client_secret["issuer"]
+            client = Client(client_authn_method=CLIENT_AUTHN_METHOD)
+            # general parameter discovery about the Identity Provider via the issuers URL
+            client.provider_config(issuer)
+            # transforming Rucio OIDC Client specific parameters into a registration form
+            client_reg = RegistrationResponse(**client_secret)
+            # storing such parameters (client_id, client_secret, etc.) into the client itself
+            client.store_registration_info(client_reg)
+            # setting public_key cache timeout to 'keytimeout' seconds
+            keybundles = client.keyjar.issuer_keys[client.issuer]
+            for keybundle in keybundles:
+                keybundle.cache_time = keytimeout
+            clients[issuer] = client
+        except:
+            continue
+    return clients
+
+
+# Initialising Rucio OIDC Clients
+OIDC_clients = get_rucio_OIDC_clients()
```

Comment:
supposed to be global to module? should be at the top of the file

Reply:
Yes. Right ! Thanks !